src/transformers/models/depth_pro/modeling_depth_pro.py

```diff
@@ -0,0 +1,1676 @@
+        if config.hidden_size % config.num_attention_heads != 0 and not hasattr(config, "embedding_size"):
+            raise ValueError(
+                f"The hidden size {config.hidden_size,} is not a multiple of the number of attention "
+                f"heads {config.num_attention_heads}."
+            )
+
+        self.num_attention_heads = config.num_attention_heads
+        self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+
+        self.query = nn.Linear(config.hidden_size, self.all_head_size, bias=config.qkv_bias)
+        self.key = nn.Linear(config.hidden_size, self.all_head_size, bias=config.qkv_bias)
+        self.value = nn.Linear(config.hidden_size, self.all_head_size, bias=config.qkv_bias)
+
+        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
+
+    # Copied from transformers.models.vit.modeling_vit.ViTSelfAttention.transpose_for_scores with ViT->DepthPro
+    def transpose_for_scores(self, x: torch.Tensor) -> torch.Tensor:
+        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
+        x = x.view(new_x_shape)
+        return x.permute(0, 2, 1, 3)
+
+    # Taken from transformers.models.vit.modeling_vit.ViTSelfAttention.forward with ViT->DepthPro
+    # with the addition of `batch_size`
+    def forward(
+        self,
+        hidden_states,
+        head_mask: Optional[torch.Tensor] = None,
+        output_attentions: bool = False,
+        batch_size: Optional[int] = None,
+    ) -> Union[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor]]:
+        mixed_query_layer = self.query(hidden_states)
+
+        key_layer = self.transpose_for_scores(self.key(hidden_states))
+        value_layer = self.transpose_for_scores(self.value(hidden_states))
+        query_layer = self.transpose_for_scores(mixed_query_layer)
+
+        # Take the dot product between "query" and "key" to get the raw attention scores.
+        attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2))
+
+        attention_scores = attention_scores / math.sqrt(self.attention_head_size)
+
+        # Normalize the attention scores to probabilities.
+        attention_probs = nn.functional.softmax(attention_scores, dim=-1)
+
+        # This is actually dropping out entire tokens to attend to, which might
+        # seem a bit unusual, but is taken from the original Transformer paper.
+        attention_probs = self.dropout(attention_probs)
+
+        # Mask heads if we want to
+        if head_mask is not None:
+            attention_probs = attention_probs * head_mask
+
+        if batch_size is not None:
+            attention_probs_batched = patch_to_batch(attention_probs, batch_size)
+            attention_probs_patched = batch_to_patch(attention_probs_batched)
+        else:
+            attention_probs_patched = attention_probs_batched = attention_probs
+
```

Comment:
@qubvel 

The problem with doing transpositions outside of the module is that the resultant vectors are outside the computation graph and their gradients are not calculated during back propagation.

I have tried to demonstrate this in a minimal example below, which mimics our problem at a lower level.

<details>

<summary>
example
</summary>

```py
import torch

# suppose these are the hidden states
x = torch.tensor([0.1, 0.2, 0.3], requires_grad=True)

# suppose these are the reshaped hidden states similar as `patch_to_batch` function
x_ = x.reshape(3, 1)

x.retain_grad()
x_.retain_grad()

x.mean().backward()

x.grad  # [0.3333, 0.3333, 0.3333]
x_.grad # None
```

</details>

Reply:
Maybe I didn't get the problem, but I meant to do it outside the Attention module, but still inside the model, I mean instead of passing hidden_state and applying path_to_batch, batch_to_patch inside we can first apply transpositions and then pass hidden states to Attention

Reply:
I can propose 2 solutions that by-passes this problem
- we can return hidden states from `image_encoder` instead of `patch_encoder`
  - both are vision transformers used in encoder
  - `patch_encoder`
    - takes the patches from images of different scales
    - input vector is of shape `num_patches, channels, patch_size, patch_size`
    - indexing is applied to match output hidden states to length batch size, which creates a new computation graph
  - `image_encoder`
    - takes original image scaled to patch_size, thus takes that single patch/image as input
    - input vector is of shape `B, channels, patch_size, patch_size`
    - doesnot require indexing, because output hidden states are already of length batch size, no new computation graph is created
  - Although `patch_encoder` seems to be more significant than `image_encoder`, yet using states from `image_encoder` can keep the code simple by ensuring the copied part is copied as it is and no external transformations are applied on the states
- we can overload `ModelTesterMixin.test_retain_grad_hidden_states_attentions` to bypass the only failing test case, this also ensures the code is copied without any change for transformer blocks. But I am not sure if that affects any use-case.